### PR TITLE
feat: support relay-style pagination, apply on uses & invitations

### PR DESCRIPTION
**Pull request by darknight**
For closing issue: https://github.com/TabbyML/tabby/issues/1107

By referring to https://github.com/dyedgreen/juniper-relay & https://github.com/jordanwebster/swapi-juniper/blob/main/src/main.rs, I managed to have a runnable implementation for `usersNext` api.

Although `invitationsNext` is missing currently, I want to raise PR early to collect feedback about current design decision.

For **relay** mod added in `juniper-axum` crate
- the objects' implementations are quite similar to above reference
- besides, I added 2 public interface, `query` & `query_async`, each accepts 4 paging arguments + 1 callback, which is for fetching node results

For **web-server**
- the `User` object must implement `NodeType` to mark it as a node
- ~~we can not derive `GraphQLObject` macro for `User` directly, which will cause compile error due to trait bound not implemented, instead, we have to define `User` as graphql object by using `#[graphql_object]`~~ (by adding `context` attribute, issue resolved)
- the meaningful cursor for `User` should be `id`, so include it in `User`

For **usersNext** api:
- The paging fitlers (before, after, first, last) are all optional, so theoretically we can pass any combinations (except `first + last`, that's strongly discourged by the spec). Well, personally, I don't think they all make sense, for example, (first + before). For current implementation, I only take care of:
  - `(first + after)` => forward pagination
  - `(last + before)` => backward pagination
  - `default` => get all users
  - **how to handle which case can be discussed**

TODO:
- code cleanup
- add tests
- ~~add `invitationsNext` implementation~~

### CODE REVIEW

File: ee/tabby-webserver/src/service/db/users.rs [hunk 1]
```diff
@@ -163,6 +171,47 @@ impl DbConn {
...
 
+    pub async fn list_users_with_filter(
+        &self,
+        limit: Option<usize>,
+        skip_id: Option<i32>,
+        backwards: bool,
+    ) -> Result<Vec<User>> {
+        let mut source = String::new();
+        let mut clause = String::new();
+        if backwards {
```
[the remark this answers is not in the document]
Comment (wsxiaoys) [reply ee/tabby-webserver/src/service/db/users.rs:182]: at least https://github.com/dyedgreen/juniper-relay/blob/master/src/lib.rs#L12 shows that GraphQLObject could coexists with RelayConnectionNode

File: ee/tabby-webserver/src/service/db/users.rs [hunk 2]
```diff
@@ -163,6 +171,47 @@ impl DbConn {
         Ok(users)
     }
 
+    pub async fn list_users_with_filter(
+        &self,
+        limit: Option<usize>,
+        skip_id: Option<i32>,
+        backwards: bool,
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/db/users.rs:178]: Why is`backwards` necessary?

File: ee/tabby-webserver/src/schema/mod.rs [hunk 3]
```diff
@@ -128,16 +128,81 @@ impl Query {
...
+            },
+        )
+        .await
+    }
 }
 
-#[derive(Debug, GraphQLObject)]
+#[derive(Default, Debug)]
 pub struct User {
+    pub id: i32,
```
[the remark this answers is not in the document]
Comment (darknight) [reply ee/tabby-webserver/src/schema/mod.rs:160]: the `cursor` field should be string, `node` could be any type except a `list`.

File: ee/tabby-webserver/src/schema/mod.rs [hunk 4]
```diff
@@ -128,16 +128,81 @@ impl Query {
...
+    fn connection_type_name() -> &'static str {
+        "UserConnection"
+    }
+
+    fn edge_type_name() -> &'static str {
+        "UserEdge"
+    }
+}
+
+#[graphql_object(context = Context)]
```
[the remark this answers is not in the document]
Comment (darknight) [reply ee/tabby-webserver/src/schema/mod.rs:183]: Yes, that's the error.

Actually this error trapped me for quite a while, I'm still not clear how to suppress it and keep using `#[derive(GraphQLObject)]`

I'm still looking into it.

File: ee/tabby-webserver/src/schema/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (wsxiaoys) [reply ee/tabby-webserver/src/schema/mod.rs:160]: i was referring the id type  https://relay.dev/graphql/objectidentification.htm

File: ee/tabby-webserver/src/schema/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (wsxiaoys) [reply ee/tabby-webserver/src/schema/mod.rs:183]: at least https://github.com/dyedgreen/juniper-relay/blob/master/src/lib.rs#L12 shows that GraphQLObject could coexists with RelayConnectionNode

Maybe worth to look at the diff on design choices

File: ee/tabby-webserver/src/service/db/users.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (darknight) [reply ee/tabby-webserver/src/service/db/users.rs:182]: > I feel there should be safer way to construct sql as complex as this (i haven't researched through this area sufficient, though)

Or we can split the logic into separate functions to simplify the implementation, here I just put forward/backward pagination together.

> at least https://github.com/dyedgreen/juniper-relay/blob/master/src/lib.rs#L12 shows that GraphQLObject could coexists with RelayConnectionNode

This is db query logic, it has nothing to do with GraphQLObject, right?
Actually, for user info, we've implemented `From<db::User> for schema::User`

[the remark this answers is not in the document]
Comment (wsxiaoys) [reply ee/tabby-webserver/src/service/db/users.rs:182]: not backwards specific, just wondering if there's a better way to format sql query. Did a quick search: https://github.com/SeaQL/sea-query, might worth to take a look

File: ee/tabby-webserver/src/schema/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (darknight) [reply ee/tabby-webserver/src/schema/mod.rs]: By adding `#[graphql(context = Context)]` to `User`, the error is fixed.

**Review by wsxiaoys — APPROVED** [review]
LGTM for relay style pagination

Please flip this out of draft if it's ok to merge. (assumming you still wanna add some unit test)

**Comment by darknight** [comment]
> LGTM for relay style pagination
> 
> Please flip this out of draft if it's ok to merge. (assumming you still wanna add some unit test)

Noted, let me add some tests, then I'll mark it as ready.

File: ee/tabby-webserver/src/service/db/users.rs [hunk 5]
```diff
@@ -242,4 +290,226 @@ mod tests {
         assert_eq!(user.email, new_user.email);
         assert_ne!(user.auth_token, new_user.auth_token);
     }
+
+    #[tokio::test]
+    async fn test_list_users_with_filter() {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/db/users.rs:295]: is it possible to extract `list_<type>_with_filter` into a single utility so unit test doesn't have to run it separately for users and invitations?

File: ee/tabby-webserver/src/service/db/mod.rs [hunk 6]
```diff
@@ -20,6 +20,12 @@ lazy_static! {
         AsyncMigrations::from_directory(&MIGRATIONS_DIR).unwrap();
 }
 
+pub trait Pagination: Sized {
+    fn tab_name() -> &'static str;
+    fn select_from(source: &str, clause: &str) -> String;
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/db/mod.rs:25]: I feel that the trait is a bit over design in this scenario. How about having a util function like below

```rust
fn make_pagination_query(table_name: &str, field_names: [&str], limit, cursor, backwards ...) -> String {
}
```

**Pull request merged**